{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d56769bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from tqdm import tqdm\n",
    "import json\n",
    "import requests\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "def filter_alleles(alleles: list[dict]) -> list[str]:\n",
    "    filtered_alleles = []\n",
    "    for item in alleles:\n",
    "        if item[\"confirmation_status.confirmed\"] and item[\"sequence_status.full\"]:\n",
    "            filtered_alleles.append(item[\"name\"])\n",
    "    return filtered_alleles\n",
    "\n",
    "\n",
    "def get_alleles_by_ipd_accession(ipd_accession):\n",
    "    \"\"\"\n",
    "    Retrieve a list of alleles that start with the given IPD accession.\n",
    "\n",
    "    Args:\n",
    "        ipd_accession (str): The IPD accession (e.g., \"B*08\")\n",
    "\n",
    "    Returns:\n",
    "        list: List of allele names that match the IPD accession\n",
    "    \"\"\"\n",
    "    base_url = \"https://www.ebi.ac.uk/cgi-bin/ipd/api/allele\"\n",
    "    params = {\n",
    "        \"query\": f'startsWith(name, \"{ipd_accession}\")',\n",
    "        \"fields\": \"name, sequence_status.full, confirmation_status.confirmed\",\n",
    "        \"limit\": 1000,  # Adjust limit as needed\n",
    "    }\n",
    "\n",
    "    try:\n",
    "        response = requests.get(base_url, params=params)\n",
    "        response.raise_for_status()  # Raise exception for HTTP errors\n",
    "\n",
    "        alleles = response.json().get(\"data\", [])\n",
    "        alleles = filter_alleles(alleles)\n",
    "        return alleles\n",
    "\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        print(f\"Request failed: {str(e)}\")\n",
    "        return []\n",
    "\n",
    "\n",
    "def get_protein_sequence_by_allele(allele_name):\n",
    "    \"\"\"\n",
    "    Retrieve protein sequence for a specific HLA allele from IPD-IMGT/HLA API.\n",
    "\n",
    "    Args:\n",
    "        allele_name (str): The allele name (e.g., \"B*08:02\")\n",
    "\n",
    "    Returns:\n",
    "        str: Protein sequence or error message\n",
    "    \"\"\"\n",
    "    base_url = \"https://www.ebi.ac.uk/cgi-bin/ipd/api/allele/download\"\n",
    "\n",
    "    params = {\"project\": \"HLA\", \"type\": \"protein\", \"query\": f'eq(name,\"{allele_name}\")'}\n",
    "    try:\n",
    "        response = requests.get(base_url, params=params)\n",
    "        response.raise_for_status()  # Raise exception for HTTP errors\n",
    "\n",
    "        if response.text:\n",
    "            return response.text.split()[-1]\n",
    "        else:\n",
    "            return \"No sequence found for this allele\"\n",
    "\n",
    "    except requests.exceptions.RequestException as e:\n",
    "        return f\"Request failed: {str(e)}\"\n",
    "\n",
    "\n",
    "def save_to_json(allele_sequences, filename=\"allele_sequences.json\"):\n",
    "    \"\"\"Save allele sequences to a JSON file.\"\"\"\n",
    "    with open(filename, \"w\") as json_file:\n",
    "        json.dump(allele_sequences, json_file, indent=4)\n",
    "\n",
    "\n",
    "input_alleles = [\n",
    "    \"HLA-A*02:256\",\n",
    "    \"HLA-B*35:42\",\n",
    "    \"HLA-B*57:06\",\n",
    "    \"HLA-A*02:266\",\n",
    "    \"HLA-B*37:01\",\n",
    "    \"HLA-A*08:01\",\n",
    "    \"HLA-A*24:01\",\n",
    "    \"HLA-B*12:01\",\n",
    "    \"HLA-B*07:01\",\n",
    "    \"HLA-C*14:02\",\n",
    "]\n",
    "\n",
    "\n",
    "# if __name__ == \"__main__\":\n",
    "#     paired_TCR = pd.read_csv(\"data/external/TCR/VDJB/paired_TCR.csv\")\n",
    "#     data = list(set(paired_TCR[\"MHC A\"].unique()).union(paired_TCR[\"MHC B\"].unique()))\n",
    "#     data.remove(\"B2M\")\n",
    "#     ipd_accessions = list(map(lambda x: x.split(\"HLA-\")[1], data))\n",
    "\n",
    "#     allele_sequences = {}\n",
    "\n",
    "#     for ipd_accession in tqdm(ipd_accessions, total=len(ipd_accessions)):\n",
    "#         alleles = get_alleles_by_ipd_accession(ipd_accession)\n",
    "#         if alleles:\n",
    "#             allele_sequences[\"HLA-\" + ipd_accession] = {}\n",
    "\n",
    "#         for allele in alleles:\n",
    "#             protein_sequence = get_protein_sequence_by_allele(allele)\n",
    "\n",
    "#             if (\n",
    "#                 \"Request failed\" not in protein_sequence\n",
    "#                 and \"No sequence found\" not in protein_sequence\n",
    "#             ):\n",
    "#                 allele_sequences[\"HLA-\" + ipd_accession][allele] = protein_sequence\n",
    "\n",
    "#     for key in allele_sequences:\n",
    "#         if len(allele_sequences[key]) == 0:\n",
    "#             del allele_sequences[key]\n",
    "\n",
    "#     save_to_json(allele_sequences, \"/home/user11/data/mhcii/allele_sequences.json\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "01196cb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "def format_name(name):\n",
    "    if name == '':\n",
    "        return ''\n",
    "    elif name == 'HLA-DRA0101':\n",
    "        return 'DRA*01:01:01:01'\n",
    "    else:\n",
    "        _, name = name.split('-')\n",
    "        first_part = name[:4]\n",
    "        second_part = name[4:6]\n",
    "        third_part = name[6:8]\n",
    "        return f'{first_part}*{second_part}:{third_part}'\n",
    "\n",
    "\n",
    "path = '/home/user11/data/data_processed/data.tsv'\n",
    "df = pd.read_csv(path, sep='\\t', header=None)\n",
    "counter = Counter(df[2]).most_common()\n",
    "ids = set(df[2])\n",
    "\n",
    "ids_table = []\n",
    "for i in ids:\n",
    "    line = dict()\n",
    "    if not i.startswith('HLA-'):\n",
    "        if 'A' in i:\n",
    "            a_chain = f\"{i.replace('_', '')}\"\n",
    "            b_chain = f'{a_chain[:2]}B'\n",
    "        if 'B' in i:\n",
    "            b_chain = f\"{i.replace('_', '')}\"\n",
    "            a_chain = f'{b_chain[:2]}A0101'\n",
    "    else:\n",
    "        _, a_chain, b_chain = i.split('-')\n",
    "    line['name_A'] = f\"HLA-{a_chain}\"\n",
    "    line['name_B'] = f\"HLA-{b_chain}\"\n",
    "    ids_table.append(line)\n",
    "\n",
    "table = pd.DataFrame(ids_table)\n",
    "table['allele_A'] = table['name_A'].apply(format_name)\n",
    "table['allele_B'] = table['name_B'].apply(format_name)\n",
    "ipd_accessions = list(set(table[\"allele_A\"].unique()).union(table[\"allele_B\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "12ebc123",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('DRB1_0101', 10280),\n",
       " ('DRB1_0701', 6237),\n",
       " ('DRB1_0401', 6188),\n",
       " ('DRB1_1101', 5900),\n",
       " ('DRB1_0301', 5223),\n",
       " ('DRB5_0101', 5053),\n",
       " ('DRB1_1501', 4766),\n",
       " ('DRB3_0101', 4585),\n",
       " ('DRB1_1302', 4440),\n",
       " ('DRB1_0802', 4433),\n",
       " ('DRB1_0901', 4280),\n",
       " ('DRB1_0405', 3923),\n",
       " ('DRB4_0101', 3903),\n",
       " ('DRB1_0404', 3601),\n",
       " ('HLA-DQA10501-DQB10301', 3546),\n",
       " ('DRB3_0202', 3315),\n",
       " ('HLA-DQA10301-DQB10302', 3063),\n",
       " ('HLA-DQA10101-DQB10501', 2913),\n",
       " ('HLA-DQA10401-DQB10402', 2858),\n",
       " ('HLA-DQA10501-DQB10201', 2778),\n",
       " ('HLA-DPA10103-DPB10401', 2696),\n",
       " ('HLA-DQA10102-DQB10602', 2676),\n",
       " ('HLA-DPA10301-DPB10402', 2611),\n",
       " ('HLA-DPA10201-DPB10501', 2440),\n",
       " ('HLA-DPA10201-DPB10101', 2417),\n",
       " ('DRB1_1201', 2360),\n",
       " ('HLA-DPA10201-DPB11401', 2285),\n",
       " ('DRB1_1001', 2053),\n",
       " ('DRB1_1602', 1686),\n",
       " ('HLA-DPA10103-DPB10301', 1550),\n",
       " ('DRB1_1301', 1007),\n",
       " ('HLA-DQA10201-DQB10202', 936),\n",
       " ('DRB1_0801', 931),\n",
       " ('HLA-DQA10104-DQB10503', 877),\n",
       " ('DRB3_0301', 877),\n",
       " ('DRB4_0103', 840),\n",
       " ('HLA-DQA10501-DQB10302', 838),\n",
       " ('HLA-DQA10102-DQB10501', 825),\n",
       " ('HLA-DQA10201-DQB10301', 818),\n",
       " ('HLA-DQA10102-DQB10502', 794),\n",
       " ('HLA-DPA10103-DPB10201', 787),\n",
       " ('HLA-DQA10201-DQB10402', 765),\n",
       " ('HLA-DQA10201-DQB10303', 759),\n",
       " ('HLA-DQA10501-DQB10402', 747),\n",
       " ('HLA-DPA10103-DPB10601', 584),\n",
       " ('HLA-DQA10303-DQB10402', 567),\n",
       " ('HLA-DQA10601-DQB10402', 565),\n",
       " ('HLA-DQA10501-DQB10303', 564),\n",
       " ('HLA-DQA10103-DQB10603', 446),\n",
       " ('HLA-DQA10301-DQB10301', 188),\n",
       " ('DRB1_0403', 59),\n",
       " ('HLA-DPA10103-DPB10402', 45),\n",
       " ('DRB1_0402', 42),\n",
       " ('DRB1_0103', 40),\n",
       " ('DRB1_0302', 37),\n",
       " ('HLA-DQA10302-DQB10401', 27),\n",
       " ('DRB1_1104', 27),\n",
       " ('DRB1_1502', 22),\n",
       " ('HLA-DQA10201-DQB10201', 19),\n",
       " ('DRB1_0406', 12),\n",
       " ('DRB1_0102', 8),\n",
       " ('DRB1_0803', 7),\n",
       " ('HLA-DQA10103-DQB10302', 6),\n",
       " ('HLA-DQA10302-DQB10303', 6),\n",
       " ('HLA-DQA10301-DQB10201', 4),\n",
       " ('DRB1_0804', 3),\n",
       " ('DRB1_0411', 2),\n",
       " ('DRB5_0102', 2),\n",
       " ('HLA-DQA10102-DQB10604', 1),\n",
       " ('HLA-DQA10505-DQB10301', 1),\n",
       " ('DRB1_1402', 1),\n",
       " ('DRB1_1503', 1)]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "151228a1",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "list indices must be integers or slices, not str",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[3]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mcounter\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mHLA-DRB1*04:02\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n",
      "\u001b[31mTypeError\u001b[39m: list indices must be integers or slices, not str"
     ]
    }
   ],
   "source": [
    "counter['HLA-DRB1*04:02']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2b73ed0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 73/73 [08:52<00:00,  7.29s/it]\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "ipd_accessions = list(set(table[\"allele_A\"].unique()).union(table[\"allele_B\"].unique()))\n",
    "\n",
    "allele_sequences = defaultdict(dict)\n",
    "\n",
    "index = 0 \n",
    "for ipd_accession in tqdm(ipd_accessions, total=len(ipd_accessions)):\n",
    "    alleles = get_alleles_by_ipd_accession(ipd_accession)\n",
    "\n",
    "    for allele in alleles:\n",
    "        protein_sequence = get_protein_sequence_by_allele(allele)\n",
    "\n",
    "        if (\n",
    "            \"Request failed\" not in protein_sequence\n",
    "            and \"No sequence found\" not in protein_sequence\n",
    "        ):\n",
    "            allele_sequences[\"HLA-\" + ipd_accession][allele] = protein_sequence\n",
    "\n",
    "    for key in allele_sequences:\n",
    "        if len(allele_sequences[key]) == 0:\n",
    "            del allele_sequences[key]\n",
    "\n",
    "save_to_json(allele_sequences, \"/home/user11/data/mhcii/allele_sequences.json\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcec4c2b",
   "metadata": {},
   "source": [
    "## Для тестовой выборки"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f46686da",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "def format_name(name):\n",
    "    if name == '':\n",
    "        return ''\n",
    "    elif name == 'HLA-DRA0101':\n",
    "        return 'DRA*01:01:01:01'\n",
    "    else:\n",
    "        _, name = name.split('-')\n",
    "        first_part = name[:4]\n",
    "        second_part = name[4:6]\n",
    "        third_part = name[6:8]\n",
    "        return f'{first_part}*{second_part}:{third_part}'\n",
    "\n",
    "\n",
    "path = '/home/user11/data/embeddings_test/test_dataset.csv'\n",
    "df = pd.read_csv(path, sep='\\t')\n",
    "counter = Counter(df['allele']).most_common()\n",
    "ids = set(df['allele'])\n",
    "\n",
    "ids_table = []\n",
    "for i in ids:\n",
    "    line = dict()\n",
    "    if not i.startswith('HLA-'):\n",
    "        if 'A' in i:\n",
    "            a_chain = f\"{i.replace('_', '')}\"\n",
    "            b_chain = f'{a_chain[:2]}B'\n",
    "        if 'B' in i:\n",
    "            b_chain = f\"{i.replace('_', '')}\"\n",
    "            a_chain = f'{b_chain[:2]}A0101'\n",
    "    else:\n",
    "        _, a_chain, b_chain = i.split('-')\n",
    "    line['name_A'] = f\"HLA-{a_chain}\"\n",
    "    line['name_B'] = f\"HLA-{b_chain}\"\n",
    "    ids_table.append(line)\n",
    "\n",
    "table = pd.DataFrame(ids_table)\n",
    "# table['allele_A'] = table['name_A'].apply(format_name)\n",
    "# table['allele_B'] = table['name_B'].apply(format_name)\n",
    "table['allele_A'] = [i.split('-')[1] for i in table['name_A']]\n",
    "table['allele_B'] = [i.split('-')[1] for i in table['name_B']]\n",
    "\n",
    "ipd_accessions = list(set(table[\"allele_A\"].unique()).union(table[\"allele_B\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f5bb732d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name_A</th>\n",
       "      <th>name_B</th>\n",
       "      <th>allele_A</th>\n",
       "      <th>allele_B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*10:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*10:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HLA-DQA1*01:02</td>\n",
       "      <td>HLA-DQB1*04:02</td>\n",
       "      <td>DQA1*01:02</td>\n",
       "      <td>DQB1*04:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HLA-DPA1*02:01</td>\n",
       "      <td>HLA-DPB1*05:01</td>\n",
       "      <td>DPA1*02:01</td>\n",
       "      <td>DPB1*05:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*11:04</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*11:04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB3*02:02</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB3*02:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*15:02</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*15:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HLA-DPA1*01:03</td>\n",
       "      <td>HLA-DPB1*02:01</td>\n",
       "      <td>DPA1*01:03</td>\n",
       "      <td>DPB1*02:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*04:04</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*04:04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-DPA1*01:03</td>\n",
       "      <td>HLA-DPB1*03:01</td>\n",
       "      <td>DPA1*01:03</td>\n",
       "      <td>DPB1*03:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-DPA1*02:01</td>\n",
       "      <td>HLA-DPB1*14:01</td>\n",
       "      <td>DPA1*02:01</td>\n",
       "      <td>DPB1*14:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*13:02</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*13:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HLA-DQA1*01:02</td>\n",
       "      <td>HLA-DQB1*02:01</td>\n",
       "      <td>DQA1*01:02</td>\n",
       "      <td>DQB1*02:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*11:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*11:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HLA-DQA1*01:02</td>\n",
       "      <td>HLA-DQB1*05:01</td>\n",
       "      <td>DQA1*01:02</td>\n",
       "      <td>DQB1*05:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HLA-DQA1*01:02</td>\n",
       "      <td>HLA-DQB1*03:01</td>\n",
       "      <td>DQA1*01:02</td>\n",
       "      <td>DQB1*03:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*15:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*15:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HLA-DQA1*01:02</td>\n",
       "      <td>HLA-DQB1*03:02</td>\n",
       "      <td>DQA1*01:02</td>\n",
       "      <td>DQB1*03:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*04:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*04:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HLA-DQA1*01:01</td>\n",
       "      <td>HLA-DQB1*05:01</td>\n",
       "      <td>DQA1*01:01</td>\n",
       "      <td>DQB1*05:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB3*01:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB3*01:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB3*01:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB3*01:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB5*01:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB5*01:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*04:05</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*04:05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*01:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*01:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*08:02</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*08:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HLA-H-2-IAb</td>\n",
       "      <td>HLA-H-B</td>\n",
       "      <td>H</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HLA-DPA1*01:03</td>\n",
       "      <td>HLA-DPB1*04:01</td>\n",
       "      <td>DPA1*01:03</td>\n",
       "      <td>DPB1*04:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB4*01:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB4*01:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*04:02</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*04:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*12:02</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*12:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*09:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*09:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*12:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*12:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HLA-DQA1*03:01</td>\n",
       "      <td>HLA-DQB1*03:02</td>\n",
       "      <td>DQA1*03:01</td>\n",
       "      <td>DQB1*03:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HLA-DQA1*05:01</td>\n",
       "      <td>HLA-DQB1*03:01</td>\n",
       "      <td>DQA1*05:01</td>\n",
       "      <td>DQB1*03:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>HLA-DPA1*01:03</td>\n",
       "      <td>HLA-DPB1*14:01</td>\n",
       "      <td>DPA1*01:03</td>\n",
       "      <td>DPB1*14:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*16:02</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*16:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*08:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*08:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*03:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*03:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>HLA-DQA1*01:02</td>\n",
       "      <td>HLA-DQB1*06:02</td>\n",
       "      <td>DQA1*01:02</td>\n",
       "      <td>DQB1*06:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>HLA-DPA1*01:03</td>\n",
       "      <td>HLA-DPB1*05:01</td>\n",
       "      <td>DPA1*01:03</td>\n",
       "      <td>DPB1*05:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>HLA-H-2-IAk</td>\n",
       "      <td>HLA-H-B</td>\n",
       "      <td>H</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*13:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*13:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*14:02</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*14:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>HLA-DPA1*03:01</td>\n",
       "      <td>HLA-DPB1*04:02</td>\n",
       "      <td>DPA1*03:01</td>\n",
       "      <td>DPB1*04:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>HLA-DPA1*01:03</td>\n",
       "      <td>HLA-DPB1*04:02</td>\n",
       "      <td>DPA1*01:03</td>\n",
       "      <td>DPB1*04:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>HLA-DRA1*01:01</td>\n",
       "      <td>HLA-DRB1*07:01</td>\n",
       "      <td>DRA1*01:01</td>\n",
       "      <td>DRB1*07:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>HLA-DQA1*05:01</td>\n",
       "      <td>HLA-DQB1*02:01</td>\n",
       "      <td>DQA1*05:01</td>\n",
       "      <td>DQB1*02:01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            name_A          name_B    allele_A    allele_B\n",
       "0   HLA-DRA1*01:01  HLA-DRB1*10:01  DRA1*01:01  DRB1*10:01\n",
       "1   HLA-DQA1*01:02  HLA-DQB1*04:02  DQA1*01:02  DQB1*04:02\n",
       "2   HLA-DPA1*02:01  HLA-DPB1*05:01  DPA1*02:01  DPB1*05:01\n",
       "3   HLA-DRA1*01:01  HLA-DRB1*11:04  DRA1*01:01  DRB1*11:04\n",
       "4   HLA-DRA1*01:01  HLA-DRB3*02:02  DRA1*01:01  DRB3*02:02\n",
       "5   HLA-DRA1*01:01  HLA-DRB1*15:02  DRA1*01:01  DRB1*15:02\n",
       "6   HLA-DPA1*01:03  HLA-DPB1*02:01  DPA1*01:03  DPB1*02:01\n",
       "7   HLA-DRA1*01:01  HLA-DRB1*04:04  DRA1*01:01  DRB1*04:04\n",
       "8   HLA-DPA1*01:03  HLA-DPB1*03:01  DPA1*01:03  DPB1*03:01\n",
       "9   HLA-DPA1*02:01  HLA-DPB1*14:01  DPA1*02:01  DPB1*14:01\n",
       "10  HLA-DRA1*01:01  HLA-DRB1*13:02  DRA1*01:01  DRB1*13:02\n",
       "11  HLA-DQA1*01:02  HLA-DQB1*02:01  DQA1*01:02  DQB1*02:01\n",
       "12  HLA-DRA1*01:01  HLA-DRB1*11:01  DRA1*01:01  DRB1*11:01\n",
       "13  HLA-DQA1*01:02  HLA-DQB1*05:01  DQA1*01:02  DQB1*05:01\n",
       "14  HLA-DQA1*01:02  HLA-DQB1*03:01  DQA1*01:02  DQB1*03:01\n",
       "15  HLA-DRA1*01:01  HLA-DRB1*15:01  DRA1*01:01  DRB1*15:01\n",
       "16  HLA-DQA1*01:02  HLA-DQB1*03:02  DQA1*01:02  DQB1*03:02\n",
       "17  HLA-DRA1*01:01  HLA-DRB1*04:01  DRA1*01:01  DRB1*04:01\n",
       "18  HLA-DQA1*01:01  HLA-DQB1*05:01  DQA1*01:01  DQB1*05:01\n",
       "19  HLA-DRA1*01:01  HLA-DRB3*01:01  DRA1*01:01  DRB3*01:01\n",
       "20  HLA-DRA1*01:01  HLA-DRB3*01:01  DRA1*01:01  DRB3*01:01\n",
       "21  HLA-DRA1*01:01  HLA-DRB5*01:01  DRA1*01:01  DRB5*01:01\n",
       "22  HLA-DRA1*01:01  HLA-DRB1*04:05  DRA1*01:01  DRB1*04:05\n",
       "23  HLA-DRA1*01:01  HLA-DRB1*01:01  DRA1*01:01  DRB1*01:01\n",
       "24  HLA-DRA1*01:01  HLA-DRB1*08:02  DRA1*01:01  DRB1*08:02\n",
       "25     HLA-H-2-IAb         HLA-H-B           H           H\n",
       "26  HLA-DPA1*01:03  HLA-DPB1*04:01  DPA1*01:03  DPB1*04:01\n",
       "27  HLA-DRA1*01:01  HLA-DRB4*01:01  DRA1*01:01  DRB4*01:01\n",
       "28  HLA-DRA1*01:01  HLA-DRB1*04:02  DRA1*01:01  DRB1*04:02\n",
       "29  HLA-DRA1*01:01  HLA-DRB1*12:02  DRA1*01:01  DRB1*12:02\n",
       "30  HLA-DRA1*01:01  HLA-DRB1*09:01  DRA1*01:01  DRB1*09:01\n",
       "31  HLA-DRA1*01:01  HLA-DRB1*12:01  DRA1*01:01  DRB1*12:01\n",
       "32  HLA-DQA1*03:01  HLA-DQB1*03:02  DQA1*03:01  DQB1*03:02\n",
       "33  HLA-DQA1*05:01  HLA-DQB1*03:01  DQA1*05:01  DQB1*03:01\n",
       "34  HLA-DPA1*01:03  HLA-DPB1*14:01  DPA1*01:03  DPB1*14:01\n",
       "35  HLA-DRA1*01:01  HLA-DRB1*16:02  DRA1*01:01  DRB1*16:02\n",
       "36  HLA-DRA1*01:01  HLA-DRB1*08:01  DRA1*01:01  DRB1*08:01\n",
       "37  HLA-DRA1*01:01  HLA-DRB1*03:01  DRA1*01:01  DRB1*03:01\n",
       "38  HLA-DQA1*01:02  HLA-DQB1*06:02  DQA1*01:02  DQB1*06:02\n",
       "39  HLA-DPA1*01:03  HLA-DPB1*05:01  DPA1*01:03  DPB1*05:01\n",
       "40     HLA-H-2-IAk         HLA-H-B           H           H\n",
       "41  HLA-DRA1*01:01  HLA-DRB1*13:01  DRA1*01:01  DRB1*13:01\n",
       "42  HLA-DRA1*01:01  HLA-DRB1*14:02  DRA1*01:01  DRB1*14:02\n",
       "43  HLA-DPA1*03:01  HLA-DPB1*04:02  DPA1*03:01  DPB1*04:02\n",
       "44  HLA-DPA1*01:03  HLA-DPB1*04:02  DPA1*01:03  DPB1*04:02\n",
       "45  HLA-DRA1*01:01  HLA-DRB1*07:01  DRA1*01:01  DRB1*07:01\n",
       "46  HLA-DQA1*05:01  HLA-DQB1*02:01  DQA1*05:01  DQB1*02:01"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "238f5834",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████| 46/46 [07:18<00:00,  9.52s/it]\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "ipd_accessions = list(set(table[\"allele_A\"].unique()).union(table[\"allele_B\"].unique()))\n",
    "\n",
    "allele_sequences = defaultdict(dict)\n",
    "\n",
    "index = 0 \n",
    "for ipd_accession in tqdm(ipd_accessions, total=len(ipd_accessions)):\n",
    "    if 'DRA1*01:01'== ipd_accession:\n",
    "        alleles = get_alleles_by_ipd_accession('DRA*01:01:01:01')\n",
    "    else:\n",
    "        alleles = get_alleles_by_ipd_accession(ipd_accession)\n",
    "\n",
    "    for allele in alleles:\n",
    "        protein_sequence = get_protein_sequence_by_allele(allele)\n",
    "\n",
    "        if (\n",
    "            \"Request failed\" not in protein_sequence\n",
    "            and \"No sequence found\" not in protein_sequence\n",
    "        ):\n",
    "            allele_sequences[\"HLA-\" + ipd_accession][allele] = protein_sequence\n",
    "\n",
    "    for key in allele_sequences:\n",
    "        if len(allele_sequences[key]) == 0:\n",
    "            del allele_sequences[key]\n",
    "\n",
    "save_to_json(allele_sequences, \"/home/user11/data/mhcii/allele_sequences_test.json\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "6057c8da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'DRB1*08:02'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ipd_accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7f60e4cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MAISGVPVLGFFIIAVLMSAQESWAIKEEHVIIQAEFYLNPDQSGEFMFDFDGDEIFHVDMAKKETVWRLEEFGRFASFEAQGALANIAVDKANLEIMTKRSNYTPITNVPPEVTVLTNSPVELREPNVLICFIDKFTPPVVNVTWLRNGKPVTTGVSETVFLPREDHLFRKFHYLPFLPSTEDVYDCRVEHWGLDEPLLKHWEFDAPSPLPETTENVVCALGLTVGLVGIIIGTIFIIKGVRKSNAAERRGPL'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_protein_sequence_by_allele('DRA*01:01:01:01')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "80fefe95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HLA-DRB1*07:01\n",
      "HLA-DRB1*01:01\n",
      "HLA-DQB1*04:02\n",
      "HLA-DRB1*10:01\n",
      "HLA-DRB1*11:01\n",
      "HLA-DRB1*14:02\n",
      "HLA-DRB1*13:01\n",
      "HLA-DPB1*05:01\n",
      "HLA-DRB1*04:02\n",
      "HLA-DPB1*02:01\n",
      "HLA-DPB1*03:01\n",
      "HLA-DRB1*12:01\n",
      "HLA-DQB1*03:01\n",
      "HLA-DRB3*02:02\n",
      "HLA-DPB1*04:01\n",
      "HLA-DQB1*02:01\n",
      "HLA-DQA1*05:01\n",
      "HLA-DRB5*01:01\n",
      "HLA-DRB1*08:01\n",
      "HLA-DPA1*02:01\n",
      "HLA-DRB1*04:05\n",
      "HLA-DRB1*11:04\n",
      "HLA-DRB1*12:02\n",
      "HLA-DQA1*03:01\n",
      "HLA-DRB1*16:02\n",
      "HLA-DPA1*01:03\n",
      "HLA-DRB1*15:01\n",
      "HLA-DRB1*04:04\n",
      "HLA-DRB4*01:01\n",
      "HLA-H\n",
      "HLA-DRB1*13:02\n",
      "HLA-DQA1*01:01\n",
      "HLA-DQB1*05:01\n",
      "HLA-DRB3*01:01\n",
      "HLA-DPA1*03:01\n",
      "HLA-DPB1*14:01\n",
      "HLA-DRB1*15:02\n",
      "HLA-DRB1*04:01\n",
      "HLA-DRB1*03:01\n",
      "HLA-DQB1*03:02\n",
      "HLA-DQA1*01:02\n",
      "HLA-DRB1*09:01\n",
      "HLA-DQB1*06:02\n",
      "HLA-DPB1*04:02\n",
      "HLA-DRB1*08:02\n"
     ]
    }
   ],
   "source": [
    "for i in allele_sequences.keys():\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd4d565d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'DRB1*01:01:01:01': 'MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVRLLERCIYNQEESVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRRVEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS',\n",
       " 'DRB1*01:01:02': 'MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVRLLERCIYNQEESVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRRVEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS',\n",
       " 'DRB1*01:01:07': 'MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVRLLERCIYNQEESVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRRVEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS',\n",
       " 'DRB1*01:01:01:02': 'MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVRLLERCIYNQEESVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRRVEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS',\n",
       " 'DRB1*01:01:39': 'MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVRLLERCIYNQEESVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRRVEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS',\n",
       " 'DRB1*01:01:41': 'MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVRLLERCIYNQEESVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRRVEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS',\n",
       " 'DRB1*01:01:01:19': 'MVCLKLPGGSCMTALTVTLMVLSSPLALAGDTRPRFLWQLKFECHFFNGTERVRLLERCIYNQEESVRFDSDVGEYRAVTELGRPDAEYWNSQKDLLEQRRAAVDTYCRHNYGVGESFTVQRRVEPKVTVYPSKTQPLQHHNLLVCSVSGFYPGSIEVRWFRNGQEEKAGVVSTGLIQNGDWTFQTLVMLETVPRSGEVYTCQVEHPSVTSPLTVEWRARSESAQSKMLSGVGGFVLGLLFLGAGLFIYFRNQKGHSGLQPTGFLS'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# QEFFIASGAAVDAIM WLFLECYDLQRATYHVGFT\n",
    "# QEFFIASGAAVDAIM\n",
    "# 9 11 22 24 31 52 53 58 59 61 65 66 68 72 73\n",
    "\n",
    "allele_sequences['HLA-DRB1*01:01']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "950d9fb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "# # Путь к исполняемому файлу ClustalW\n",
    "# clustalw_exe = \"clustalw2\"  # Укажите путь, если требуется\n",
    "\n",
    "# # Входной файл с последовательностями\n",
    "# input_file = \"hla_sequences.fasta\"\n",
    "# output_file = \"aligned_hla.aln\"\n",
    "\n",
    "# # Создание команды ClustalW\n",
    "# clustalw_cline = ClustalwCommandline(clustalw_exe, infile=input_file, outfile=output_file, align=True)\n",
    "\n",
    "# # Выполнение выравнивания\n",
    "# stdout, stderr = clustalw_cline()\n",
    "\n",
    "# # Чтение выровненных последовательностей\n",
    "# alignment = AlignIO.read(output_file, \"clustal\")\n",
    "\n",
    "# # Вывод выравнивания\n",
    "# print(alignment)\n",
    "\n",
    "# # Анализ позиций\n",
    "# for record in alignment:\n",
    "#     print(f\"ID: {record.id}, Sequence: {record.seq}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "6d352d01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MultiIndex([( 'Receptor',             'Group IRI'),\n",
       "            ( 'Receptor',      'IEDB Receptor ID'),\n",
       "            ( 'Receptor',        'Reference Name'),\n",
       "            ( 'Receptor',                  'Type'),\n",
       "            ('Reference',              'IEDB IRI'),\n",
       "            (  'Epitope',              'IEDB IRI'),\n",
       "            (  'Epitope',                  'Name'),\n",
       "            (  'Epitope',       'Source Molecule'),\n",
       "            (  'Epitope',       'Source Organism'),\n",
       "            (    'Assay',                  'Type'),\n",
       "            (    'Assay',              'IEDB IDs'),\n",
       "            (    'Assay',      'MHC Allele Names'),\n",
       "            (  'Chain 1',                  'Type'),\n",
       "            (  'Chain 1',          'Organism IRI'),\n",
       "            (  'Chain 1',   'Nucleotide Sequence'),\n",
       "            (  'Chain 1',        'Curated V Gene'),\n",
       "            (  'Chain 1',     'Calculated V Gene'),\n",
       "            (  'Chain 1',        'Curated D Gene'),\n",
       "            (  'Chain 1',     'Calculated D Gene'),\n",
       "            (  'Chain 1',        'Curated J Gene'),\n",
       "            (  'Chain 1',     'Calculated J Gene'),\n",
       "            (  'Chain 1',      'Protein Sequence'),\n",
       "            (  'Chain 1',           'Protein IRI'),\n",
       "            (  'Chain 1',          'CDR3 Curated'),\n",
       "            (  'Chain 1',       'CDR3 Calculated'),\n",
       "            (  'Chain 1',    'CDR3 Start Curated'),\n",
       "            (  'Chain 1',      'CDR3 End Curated'),\n",
       "            (  'Chain 1', 'CDR3 Start Calculated'),\n",
       "            (  'Chain 1',   'CDR3 End Calculated'),\n",
       "            (  'Chain 1',          'CDR1 Curated'),\n",
       "            (  'Chain 1',       'CDR1 Calculated'),\n",
       "            (  'Chain 1',    'CDR1 Start Curated'),\n",
       "            (  'Chain 1',      'CDR1 End Curated'),\n",
       "            (  'Chain 1', 'CDR1 Start Calculated'),\n",
       "            (  'Chain 1',   'CDR1 End Calculated'),\n",
       "            (  'Chain 1',          'CDR2 Curated'),\n",
       "            (  'Chain 1',       'CDR2 Calculated'),\n",
       "            (  'Chain 1',    'CDR2 Start Curated'),\n",
       "            (  'Chain 1',      'CDR2 End Curated'),\n",
       "            (  'Chain 1', 'CDR2 Start Calculated'),\n",
       "            (  'Chain 1',   'CDR2 End Calculated'),\n",
       "            (  'Chain 2',                  'Type'),\n",
       "            (  'Chain 2',          'Organism IRI'),\n",
       "            (  'Chain 2',   'Nucleotide Sequence'),\n",
       "            (  'Chain 2',        'Curated V Gene'),\n",
       "            (  'Chain 2',     'Calculated V Gene'),\n",
       "            (  'Chain 2',        'Curated D Gene'),\n",
       "            (  'Chain 2',     'Calculated D Gene'),\n",
       "            (  'Chain 2',        'Curated J Gene'),\n",
       "            (  'Chain 2',     'Calculated J Gene'),\n",
       "            (  'Chain 2',      'Protein Sequence'),\n",
       "            (  'Chain 2',           'Protein IRI'),\n",
       "            (  'Chain 2',          'CDR3 Curated'),\n",
       "            (  'Chain 2',       'CDR3 Calculated'),\n",
       "            (  'Chain 2',    'CDR3 Start Curated'),\n",
       "            (  'Chain 2',      'CDR3 End Curated'),\n",
       "            (  'Chain 2', 'CDR3 Start Calculated'),\n",
       "            (  'Chain 2',   'CDR3 End Calculated'),\n",
       "            (  'Chain 2',          'CDR1 Curated'),\n",
       "            (  'Chain 2',       'CDR1 Calculated'),\n",
       "            (  'Chain 2',    'CDR1 Start Curated'),\n",
       "            (  'Chain 2',      'CDR1 End Curated'),\n",
       "            (  'Chain 2', 'CDR1 Start Calculated'),\n",
       "            (  'Chain 2',   'CDR1 End Calculated'),\n",
       "            (  'Chain 2',          'CDR2 Curated'),\n",
       "            (  'Chain 2',       'CDR2 Calculated'),\n",
       "            (  'Chain 2',    'CDR2 Start Curated'),\n",
       "            (  'Chain 2',      'CDR2 End Curated'),\n",
       "            (  'Chain 2', 'CDR2 Start Calculated'),\n",
       "            (  'Chain 2',   'CDR2 End Calculated')],\n",
       "           )"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "receptors = dict()\n",
    "df = pd.read_csv('/home/user11/data/mhcii/tcr_full_v3.csv', low_memory=False, header=[0, 1])\n",
    "df = df.drop_duplicates(subset=[('Assay', 'MHC Allele Names')])\n",
    "df.columns\n",
    "# for index, line in df.iterrows():\n",
    "#     receptors[line[()]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "35d4ef16",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = df[df[('Assay', 'MHC Allele Names')] == 'HLA-DRA*01:01/DRB1*01:01'][(  'Chain 1',      'Protein Sequence')].to_list()\n",
    "\n",
    "#df.drop_duplicates(subset=[('Assay', 'MHC Allele Names')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "66819502",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = df[(  'Chain 1',      'Protein Sequence')].to_list()\n",
    "\n",
    "#df.drop_duplicates(subset=[('Assay', 'MHC Allele Names')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "5800c6f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = [s for s in seq if isinstance(s,str)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "f8d6ce86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ADVTQTPRNRITKTGKRIMLECSQTKGHDRMYWYRQDPGLGLRLIYYSFDVKDINKGEISDGYSVSRQAQAKFSLSLESAIPNQTALYFCATSDESYGYTFGSGTRLTVVEDLNKVSPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWVNGKEVHSGVCTDPQPLKEQPALNDSRYSLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRAD',\n",
       " 'ADVTQTPRNRITKTGKRIMLECSQTKGHDRMYWYRQDPGLGLRLIYYSFDVKDINKGEISDGYSVSRQAQAKFSLSLESAIPNQTALYFCATSDESYGYTFGSGTRLTVVEDLNKVSPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWVNGKEVHSGVCTDPQPLKEQPALNDSRYSLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRAD',\n",
       " 'ADVTQTPRNRITKTGKRIMLECSQTKGHDRMYWYRQDPGLGLRLIYYSFDVKDINKGEISDGYSVSRQAQAKFSLSLESAIPNQTALYFCATSDESYGYTFGSGTRLTVVEDLNKVSPPEVAVFEPSEAEISHTQKATLVCLATGFYPDHVELSWWVNGKEVHSGVCTDPQPLKEQPALNDSRYSLSSRLRVSATFWQNPRNHFRCQVQFYGLSENDEWTQDRAKPVTQIVSAEAWGRAD']"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# QEFFIASGAAVDAIM WLFLECYDLQRATYHVGFT\n",
    "# YAFFMFSGGAILNTL YGQFEYFAIEKVRVHLDVT\n",
    "# YAFFMFSGGAILNTL YGQFEYFAIEKVRVHLDVT\n",
    "# 9 11 22 24 31 52 53 58 59 61 65 66 68 72 73\n",
    "\n",
    "seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "99b48da7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'N'"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq[0][8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "004c39ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Class II, allele undetermined',\n",
       " 'Gaga-BF2*015:01',\n",
       " 'H2 class I',\n",
       " 'H2 class II',\n",
       " 'H2 class II, H2-b class I',\n",
       " 'H2 class II, H2-b class II',\n",
       " 'H2 class II, HLA-DRB1*15:01',\n",
       " 'H2-Db',\n",
       " 'H2-Db, H2-Db Y159F mutant',\n",
       " 'H2-IAb',\n",
       " 'H2-IAb I67F, R70Q, T71K mutant',\n",
       " 'H2-IAb, H2-a class II',\n",
       " 'H2-IAb, mouse CD1d',\n",
       " 'H2-IAd',\n",
       " 'H2-IAg7',\n",
       " 'H2-IAk',\n",
       " 'H2-IAq',\n",
       " 'H2-IAs',\n",
       " 'H2-IAu',\n",
       " 'H2-IEd',\n",
       " 'H2-IEd, H2-d class II',\n",
       " 'H2-IEk',\n",
       " 'H2-Kb',\n",
       " 'H2-Kb D77S, K89A mutant',\n",
       " 'H2-Kb Y22F, M23I, E24S, D30N mutant',\n",
       " 'H2-Kb, H2-Kb Y22F, M23I, E24S, D30N mutant',\n",
       " 'H2-Kb, H2-Kb Y84A mutant, H2-Kb Y84C mutant',\n",
       " 'H2-Kd',\n",
       " 'H2-Ld',\n",
       " 'H2-Q9',\n",
       " 'H2-Qa-1b',\n",
       " 'H2-b class I',\n",
       " 'H2-b class II',\n",
       " 'H2-d class II',\n",
       " 'H2-s class I',\n",
       " 'H2-u class II',\n",
       " 'HLA class I',\n",
       " 'HLA class II',\n",
       " 'HLA-A*01:01',\n",
       " 'HLA-A*02:01',\n",
       " 'HLA-A*02:01 K66A mutant',\n",
       " 'HLA-A*02:01 K66A, E63Q mutant',\n",
       " 'HLA-A*02:01 T163A mutant',\n",
       " 'HLA-A*02:01 W167A mutant',\n",
       " 'HLA-A*02:01, HLA-A*02:01 A150P mutant',\n",
       " 'HLA-A*02:01, HLA-A*02:01 A69G mutant, HLA-A*02:01 E166A mutant, HLA-A*02:01 K66A mutant, HLA-A*02:01 R65A mutant, HLA-A*02:01 T163A mutant, HLA-A*02:01 W167A mutant',\n",
       " 'HLA-A*02:01, HLA-A*02:01 K66A mutant, HLA-A*02:01 Q155A mutant, HLA-A*02:01 T163A mutant, HLA-A*02:01 W167A mutant',\n",
       " 'HLA-A*02:01, HLA-A*02:01 K66A, E63Q mutant',\n",
       " 'HLA-A*02:01, HLA-A*24:02',\n",
       " 'HLA-A*02:01, HLA-DPB1*11:01',\n",
       " 'HLA-A*02:05',\n",
       " 'HLA-A*02:06',\n",
       " 'HLA-A*02:10',\n",
       " 'HLA-A*03:01',\n",
       " 'HLA-A*03:01, HLA-A*11:01',\n",
       " 'HLA-A*11:01',\n",
       " 'HLA-A*23:01',\n",
       " 'HLA-A*24:02',\n",
       " 'HLA-A*29:01',\n",
       " 'HLA-A*29:02',\n",
       " 'HLA-A*30:01',\n",
       " 'HLA-A*30:02',\n",
       " 'HLA-A*32:01',\n",
       " 'HLA-A*68:01',\n",
       " 'HLA-A*80:01',\n",
       " 'HLA-A1',\n",
       " 'HLA-A11',\n",
       " 'HLA-A2',\n",
       " 'HLA-B*07:02',\n",
       " 'HLA-B*08:01',\n",
       " 'HLA-B*08:01, HLA-B*15:01, HLA-B*15:02',\n",
       " 'HLA-B*13:01',\n",
       " 'HLA-B*13:01, HLA-DQB1*05:01',\n",
       " 'HLA-B*15:01',\n",
       " 'HLA-B*15:01, HLA-B*15:02',\n",
       " 'HLA-B*15:01, HLA-B*15:02, HLA-B*15:21, HLA-B*15:25',\n",
       " 'HLA-B*15:02',\n",
       " 'HLA-B*15:02, HLA-B*15:21',\n",
       " 'HLA-B*15:25',\n",
       " 'HLA-B*18:01',\n",
       " 'HLA-B*27:05',\n",
       " 'HLA-B*27:05, HLA-B*27:09',\n",
       " 'HLA-B*35:01',\n",
       " 'HLA-B*35:01, HLA-B*35:08',\n",
       " 'HLA-B*35:03',\n",
       " 'HLA-B*35:08',\n",
       " 'HLA-B*35:08 Q155A mutant',\n",
       " 'HLA-B*35:08 Q65A, T69A, Q155A mutant',\n",
       " 'HLA-B*35:08, HLA-B*35:08 Q65A, T69A, Q155A mutant',\n",
       " 'HLA-B*37:01',\n",
       " 'HLA-B*38:01',\n",
       " 'HLA-B*39:01',\n",
       " 'HLA-B*40:01',\n",
       " 'HLA-B*40:02',\n",
       " 'HLA-B*41:02',\n",
       " 'HLA-B*44:02',\n",
       " 'HLA-B*44:02, HLA-B*44:03, HLA-B*44:05',\n",
       " 'HLA-B*44:02, HLA-B*44:05',\n",
       " 'HLA-B*44:03',\n",
       " 'HLA-B*44:03, HLA-B*44:08',\n",
       " 'HLA-B*44:05',\n",
       " 'HLA-B*48:01',\n",
       " 'HLA-B*50:01',\n",
       " 'HLA-B*51:01',\n",
       " 'HLA-B*53:01',\n",
       " 'HLA-B*57:01',\n",
       " 'HLA-B*57:01, HLA-B*57:03',\n",
       " 'HLA-B*57:03',\n",
       " 'HLA-B18',\n",
       " 'HLA-B27',\n",
       " 'HLA-B35',\n",
       " 'HLA-B57',\n",
       " 'HLA-B7',\n",
       " 'HLA-B8',\n",
       " 'HLA-C*01:02',\n",
       " 'HLA-C*03:03',\n",
       " 'HLA-C*03:04',\n",
       " 'HLA-C*04:01',\n",
       " 'HLA-C*05:01',\n",
       " 'HLA-C*06:02',\n",
       " 'HLA-C*07:01',\n",
       " 'HLA-C*07:02',\n",
       " 'HLA-C*08:01',\n",
       " 'HLA-C*08:02',\n",
       " 'HLA-C*12:02',\n",
       " 'HLA-C*14:02',\n",
       " 'HLA-C*15:02',\n",
       " 'HLA-Cw3',\n",
       " 'HLA-DP',\n",
       " 'HLA-DPA1*01:03/DPB1*02:01',\n",
       " 'HLA-DPA1*01:03/DPB1*02:01, HLA-DPA1*01:03/DPB1*04:02',\n",
       " 'HLA-DPA1*01:03/DPB1*03:01',\n",
       " 'HLA-DPA1*01:03/DPB1*04:01',\n",
       " 'HLA-DPA1*01:03/DPB1*04:02',\n",
       " 'HLA-DPA1*02:01/DPB1*05:01, HLA-DPA1*02:02/DPB1*05:01',\n",
       " 'HLA-DPA1*02:01/DPB1*09:01',\n",
       " 'HLA-DPA1*02:01/DPB1*13:01',\n",
       " 'HLA-DPA1*02:02/DPB1*02:01',\n",
       " 'HLA-DPA1*02:02/DPB1*05:01',\n",
       " 'HLA-DPB1*02:01',\n",
       " 'HLA-DPB1*04:01',\n",
       " 'HLA-DPw2',\n",
       " 'HLA-DPw4',\n",
       " 'HLA-DQ',\n",
       " 'HLA-DQ1',\n",
       " 'HLA-DQ2',\n",
       " 'HLA-DQ6',\n",
       " 'HLA-DQ8',\n",
       " 'HLA-DQA1*01:01/DQB1*05:01',\n",
       " 'HLA-DQA1*01:02/DQB1*05:02',\n",
       " 'HLA-DQA1*01:02/DQB1*06:02',\n",
       " 'HLA-DQA1*01:02/DQB1*06:02, HLA-DRA*01:01/DRB1*15:01, HLA-DRA*01:01/DRB5*01:01',\n",
       " 'HLA-DQA1*01:02/DQB1*06:04',\n",
       " 'HLA-DQA1*01:03/DQB1*06:01',\n",
       " 'HLA-DQA1*01:03/DQB1*06:01, HLA-DQA1*01:03/DQB1*06:03',\n",
       " 'HLA-DQA1*01:04/DQB1*05:03',\n",
       " 'HLA-DQA1*02:01/DQB1*02:02',\n",
       " 'HLA-DQA1*02:01/DQB1*02:02, HLA-DQA1*05:01/DQB1*02:01',\n",
       " 'HLA-DQA1*03:01/DQB1*02:01',\n",
       " 'HLA-DQA1*03:01/DQB1*02:01, HLA-DQA1*03:01/DQB1*03:02',\n",
       " 'HLA-DQA1*03:01/DQB1*03:02',\n",
       " 'HLA-DQA1*03:01/DQB1*03:02, HLA-DQA1*05:01/DQB1*03:02',\n",
       " 'HLA-DQA1*03:03/DQB1*03:01',\n",
       " 'HLA-DQA1*05:01/DQB1*02:01',\n",
       " 'HLA-DQA1*05:01/DQB1*02:01, HLA-DQA1*05:01/DQB1*02:02',\n",
       " 'HLA-DQA1*05:01/DQB1*02:01, HLA-DQA1*05:01/DQB1*02:02, HLA-DRB4*01:01',\n",
       " 'HLA-DQA1*05:01/DQB1*03:02',\n",
       " 'HLA-DQA1*05:01/DQB1*03:02, HLA-DRB4*01:01',\n",
       " 'HLA-DQB1*03:02',\n",
       " 'HLA-DQB1*05:01, HLA-DQB1*05:02',\n",
       " 'HLA-DQB1*06:01',\n",
       " 'HLA-DQB1*06:02',\n",
       " 'HLA-DR',\n",
       " 'HLA-DR, HLA-DRB1*04:04, HLA-DRB1*08:03, HLA-DRB1*09:01, HLA-DRB1*15:01, HLA-DRB1*15:03',\n",
       " 'HLA-DR, HLA-DRB1*08:03, HLA-DRB1*09:01, HLA-DRB1*13:02',\n",
       " 'HLA-DR1',\n",
       " 'HLA-DR11',\n",
       " 'HLA-DR15',\n",
       " 'HLA-DR2',\n",
       " 'HLA-DR3',\n",
       " 'HLA-DR4',\n",
       " 'HLA-DR7',\n",
       " 'HLA-DRA*01:01/DRB1*01:01',\n",
       " 'HLA-DRA*01:01/DRB1*01:01, HLA-DRA*01:01/DRB1*04:01',\n",
       " 'HLA-DRA*01:01/DRB1*03:01',\n",
       " 'HLA-DRA*01:01/DRB1*04:01',\n",
       " 'HLA-DRA*01:01/DRB1*07:01',\n",
       " 'HLA-DRA*01:01/DRB1*11:01',\n",
       " 'HLA-DRA*01:01/DRB1*15:01',\n",
       " 'HLA-DRA*01:01/DRB1*15:01, HLA-DRA*01:01/DRB5*01:01',\n",
       " 'HLA-DRA*01:01/DRB3*01:01',\n",
       " 'HLA-DRA*01:01/DRB3*02:02',\n",
       " 'HLA-DRA*01:01/DRB5*01:01',\n",
       " 'HLA-DRB1',\n",
       " 'HLA-DRB1*01:01',\n",
       " 'HLA-DRB1*01:01, HLA-DRB1*07:01',\n",
       " 'HLA-DRB1*01:01, HLA-DRB1*07:01, HLA-DRB1*11:01, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*01:01, HLA-DRB1*07:01, HLA-DRB1*15:02',\n",
       " 'HLA-DRB1*01:01, HLA-DRB1*07:01, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*01:01, HLA-DRB1*11:01, HLA-DRB1*15:02, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*03:01',\n",
       " 'HLA-DRB1*04:01',\n",
       " 'HLA-DRB1*04:01, HLA-DRB1*04:04',\n",
       " 'HLA-DRB1*04:01, HLA-DRB1*04:05',\n",
       " 'HLA-DRB1*04:02',\n",
       " 'HLA-DRB1*04:02, HLA-DRB1*04:05',\n",
       " 'HLA-DRB1*04:03',\n",
       " 'HLA-DRB1*04:04',\n",
       " 'HLA-DRB1*04:04, HLA-DRB1*08:03, HLA-DRB1*08:18, HLA-DRB1*09:01, HLA-DRB1*13:02, HLA-DRB1*15:01, HLA-DRB1*15:03, HLA-DRB1*16:02',\n",
       " 'HLA-DRB1*04:04, HLA-DRB1*08:03, HLA-DRB1*08:18, HLA-DRB1*09:01, HLA-DRB1*15:01, HLA-DRB1*15:03, HLA-DRB1*16:02',\n",
       " 'HLA-DRB1*04:04, HLA-DRB1*08:18, HLA-DRB1*09:01, HLA-DRB1*13:02, HLA-DRB1*15:01, HLA-DRB1*15:03, HLA-DRB1*16:02',\n",
       " 'HLA-DRB1*04:05',\n",
       " 'HLA-DRB1*07:01',\n",
       " 'HLA-DRB1*07:01, HLA-DRB1*11:01, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*07:01, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*08:03',\n",
       " 'HLA-DRB1*08:03, HLA-DRB1*08:18, HLA-DRB1*09:01, HLA-DRB1*13:02, HLA-DRB1*16:02',\n",
       " 'HLA-DRB1*08:03, HLA-DRB1*08:18, HLA-DRB1*13:02',\n",
       " 'HLA-DRB1*08:03, HLA-DRB1*09:01',\n",
       " 'HLA-DRB1*08:18',\n",
       " 'HLA-DRB1*08:18, HLA-DRB1*15:01',\n",
       " 'HLA-DRB1*08:18, HLA-DRB1*15:01, HLA-DRB1*15:03',\n",
       " 'HLA-DRB1*09:01',\n",
       " 'HLA-DRB1*11:01',\n",
       " 'HLA-DRB1*11:01, HLA-DRB1*15:02, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*11:01, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*12:01',\n",
       " 'HLA-DRB1*13:01',\n",
       " 'HLA-DRB1*13:01, HLA-DRB1*13:02',\n",
       " 'HLA-DRB1*13:02',\n",
       " 'HLA-DRB1*13:02, HLA-DRB1*15:03',\n",
       " 'HLA-DRB1*14:02',\n",
       " 'HLA-DRB1*14:05',\n",
       " 'HLA-DRB1*14:06',\n",
       " 'HLA-DRB1*15:01',\n",
       " 'HLA-DRB1*15:01, HLA-DRB1*15:02, HLA-DRB1*15:03, HLA-DRB1*15:04',\n",
       " 'HLA-DRB1*15:01, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*15:02',\n",
       " 'HLA-DRB1*15:02, HLA-DRB5*01:01',\n",
       " 'HLA-DRB1*15:03',\n",
       " 'HLA-DRB3*01:01',\n",
       " 'HLA-DRB3*02:01, HLA-DRB3*03:01',\n",
       " 'HLA-DRB3*02:02',\n",
       " 'HLA-DRB3*03:01',\n",
       " 'HLA-DRB4*01:01',\n",
       " 'HLA-DRB4*01:03',\n",
       " 'HLA-DRB5*01:01',\n",
       " 'HLA-DRB5*01:01, HLA-DRB5*01:02',\n",
       " 'HLA-E*01:01, HLA-E*01:03',\n",
       " 'HLA-E*01:03',\n",
       " 'Mamu-A1*065:01',\n",
       " 'cattle MR1',\n",
       " 'human CD1a',\n",
       " 'human CD1b',\n",
       " 'human CD1c',\n",
       " 'human CD1d',\n",
       " 'human CD1d, mouse CD1d',\n",
       " 'human MR1',\n",
       " 'human MR1 K43A mutant',\n",
       " 'human MR1 R9H mutant',\n",
       " 'mouse CD1d',\n",
       " 'mouse CD1d, mouse CD1d1',\n",
       " 'mouse CD1d1',\n",
       " 'mouse MR1',\n",
       " nan}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df[('Assay', 'MHC Allele Names')])\n",
    "#df[('Chain 2',           'Type')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "a7ae1d82",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      " 12%|█▏        | 9/73 [00:06<00:49,  1.29it/s]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "ipd_accessions = list(set(table[\"allele_A\"].unique()).union(table[\"allele_B\"].unique()))\n",
    "\n",
    "allele_sequences = {}\n",
    "\n",
    "index = 0 \n",
    "for ipd_accession in tqdm(ipd_accessions, total=len(ipd_accessions)):\n",
    "    alleles = get_alleles_by_ipd_accession(ipd_accession)\n",
    "    index += 1\n",
    "    if index == 10:\n",
    "        break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e0f42675",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DRA*01:01:01:01',\n",
       " 'DRA*01:01:01:02',\n",
       " 'DRA*01:01:01:03',\n",
       " 'DRA*01:01:01:05',\n",
       " 'DRA*01:01:01:04',\n",
       " 'DRA*01:01:01:10',\n",
       " 'DRA*01:01:01:11',\n",
       " 'DRA*01:01:01:08',\n",
       " 'DRA*01:01:01:07',\n",
       " 'DRA*01:01:01:06',\n",
       " 'DRA*01:01:01:12',\n",
       " 'DRA*01:01:01:16',\n",
       " 'DRA*01:01:01:13',\n",
       " 'DRA*01:01:01:14',\n",
       " 'DRA*01:01:01:15',\n",
       " 'DRA*01:01:01:18',\n",
       " 'DRA*01:01:01:20',\n",
       " 'DRA*01:01:03',\n",
       " 'DRA*01:01:01:19',\n",
       " 'DRA*01:01:01:25',\n",
       " 'DRA*01:01:01:23',\n",
       " 'DRA*01:01:01:22',\n",
       " 'DRA*01:01:01:21',\n",
       " 'DRA*01:01:01:24',\n",
       " 'DRA*01:01:04']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_alleles_by_ipd_accession('DRA*01:01')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "924715c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'D'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = 'MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEPRAPWIEQEGPEYWDQETRNMKAHSQTDRANLGTLRGYYNQSEDGSHTIQIMYGCDVGPDGRFLRGYRQDAYDGKDYIALNEDLRSWTAADMAAQITKRKWEAVHAAEQRRVYLEGRCVDGLRRYLENGKETLQRTDPPKTHMTHHPISDHEATLRCWALGFYPAEITLTWQRDGEDQTQDTELVETRPAGDGTFQKWAAVVVPSGEEQRYTCHVQHEGLPKPLTLRWELSSQPTIPIVGIIAGLVLLGAVITGAVVAAVMWRRKSSDRKGGSYTQAASSDSAQGSDVSLTACKV'\n",
    "seq[52]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "881a944e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'L'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq[8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "3348d161",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name_A</th>\n",
       "      <th>name_B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>HLA-DRB11302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HLA-DPA10103</td>\n",
       "      <td>HLA-DPB10301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td>HLA-DRB10901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td>HLA-DRB10301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HLA-DQA10303</td>\n",
       "      <td>HLA-DQB10402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>HLA-DQA10301</td>\n",
       "      <td>HLA-DQB10302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>HLA-DQA10201</td>\n",
       "      <td>HLA-DQB10303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>HLA-DQA10201</td>\n",
       "      <td>HLA-DQB10301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td></td>\n",
       "      <td>HLA-DRB11301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>HLA-DQA10301</td>\n",
       "      <td>HLA-DQB10201</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>72 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          name_A        name_B\n",
       "0                 HLA-DRB11302\n",
       "1   HLA-DPA10103  HLA-DPB10301\n",
       "2                 HLA-DRB10901\n",
       "3                 HLA-DRB10301\n",
       "4   HLA-DQA10303  HLA-DQB10402\n",
       "..           ...           ...\n",
       "67  HLA-DQA10301  HLA-DQB10302\n",
       "68  HLA-DQA10201  HLA-DQB10303\n",
       "69  HLA-DQA10201  HLA-DQB10301\n",
       "70                HLA-DRB11301\n",
       "71  HLA-DQA10301  HLA-DQB10201\n",
       "\n",
       "[72 rows x 2 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(ids_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ef02574c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import xmltodict\n",
    "\n",
    "path = '/home/user11/data/mhcii/MhcAlleleNameList.xml'\n",
    "with open(path) as file:\n",
    "    xml = file.readlines()\n",
    "\n",
    "xml_str = ''.join(xml)\n",
    "xml_dict = xmltodict.parse(xml_str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "15267115",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['@xsi:noNamespaceSchemaLocation', '@xmlns:xsi', 'MhcAlleleName'])"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xml_dict['MhcAlleleNameList'].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "59bf03d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "22697"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(xml_dict['MhcAlleleNameList']['MhcAlleleName'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "2c730bb8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'MhcAlleleRestrictionId': '425', 'DisplayedRestriction': 'HLA-DRB1*13:02', 'Synonyms': 'HLA-DRB1*1302|HLA-DRB1*130201|HLA-DRB113:02|HLA-DRB11302', 'Includes': 'HLA-DRB1*130201', 'RestrictionLevel': 'partial molecule', 'Organism': 'Homo sapiens (human)', 'OrganismNcbiTaxId': '9606', 'Class': 'II', 'Locus': 'DR', 'Molecule': 'HLA-DRB1*13:02', 'Chain1Name': 'HLA-DRA', 'Chain2Name': 'HLA-DRB1*13:02'}\n"
     ]
    }
   ],
   "source": [
    "name = 'HLA-DRB11302'\n",
    "for record in xml_dict['MhcAlleleNameList']['MhcAlleleName']:\n",
    "    # if 'DisplayedRestriction' in record:\n",
    "    #     if name == record['DisplayedRestriction']:\n",
    "    #         print(record)\n",
    "    if 'Synonyms' in record:\n",
    "        if name in record['Synonyms'].split('|'):\n",
    "            print(record)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "bd6ed6d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#allele_name = 'HLA00066'\n",
    "allele_name = 'HLA-DPA10103-DPB10601'\n",
    "container = []\n",
    "for i in ids:\n",
    "    if 'HLA-' in i:\n",
    "        allele_name = i\n",
    "    else:\n",
    "        allele_name = 'HLA-' + i\n",
    "    out = get_alleles_by_ipd_accession(allele_name)\n",
    "    container.append(out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "98bcbbb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#allele_name = 'DQA*1:05:01'\n",
    "allele_name = 'DQB1*03:02'\n",
    "out = get_alleles_by_ipd_accession(allele_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "10877686",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DQB1*03:02:01:01',\n",
       " 'DQB1*03:02:02:01',\n",
       " 'DQB1*03:02:03',\n",
       " 'DQB1*03:02:09',\n",
       " 'DQB1*03:02:11',\n",
       " 'DQB1*03:02:12',\n",
       " 'DQB1*03:02:01:02',\n",
       " 'DQB1*03:02:21',\n",
       " 'DQB1*03:02:01:04',\n",
       " 'DQB1*03:02:01:05',\n",
       " 'DQB1*03:02:32',\n",
       " 'DQB1*03:02:01:10',\n",
       " 'DQB1*03:02:01:12',\n",
       " 'DQB1*03:02:02:02',\n",
       " 'DQB1*03:02:37',\n",
       " 'DQB1*03:02:39',\n",
       " 'DQB1*03:02:01:25',\n",
       " 'DQB1*03:02:01:29']"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf484726",
   "metadata": {},
   "outputs": [],
   "source": [
    "# YAFFMFSGGAILNTL YGQFEYFAIEKVRVHLDVT\n",
    "# YAFFMFSGGAILNTL YGQFEYFAIEKVRVHLDVT\n",
    "# YAFFQFSGGAILNTL\n",
    "# 9 11 22 24 31 52 53 58 59 61 65 66 68 72 73\n",
    "# 82 - 52 = 30\n",
    "\n",
    "# zero based \n",
    "# 39, 41, 52, 54, 61, 82, 83, 88, 89, 91, 95, 96, 98, 102, 103 \n",
    "\n",
    "s1 = 'MRPEDRMFHIRAVILRALSLAFLLSLRGAGAIKADHVSTYAAFVQTHRPTGEFMFEFDEDEQFYVDLDKKETVWHLEEFGRAFSFEAQGGLANIAILNNNLNTLIQRSNHTQAANDPPEVTVFPKEPVELGQPNTLICHIDRFFPPVLNVTWLCNGEPVTEGVAESLFLPRTDYSFHKFHYLTFVPSAEDVYDCRVEHWGLDQPLLKHWEAQEPIQMPETTETVLCALGLVLGLVGIIVGTVLIIKSLRSGHDPRAQGPL'\n",
    "s2 = 'MRPEDRMFHIRAVILRALSLAFLLSLRGAGAIKADHVSTYAAFVQTHRPTGEFMFEFDEDEQFYVDLDKKETVWHLEEFGRAFSFEAQGGLANIAILNNNLNTLIQRSNHTQAANDPPEVTVFPKEPVELGQPNTLICHIDRFFPPVLNVTWLCNGEPVTEGVAESLFLPRTDYSFHKFHYLTFVPSAEDVYDCRVEHWGLDQPLLKHWEAQEPIQMPETTETVLCALGLVLGLVGIIVGTVLIIKSLRSGHDPRAQGPL'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "405c8abe",
   "metadata": {},
   "outputs": [],
   "source": [
    "'MAISGVPVLGFFIIAVLMSAQESWAIKEEHVIIQAEFYLNPDQSGEFMFDFDGDEIFHVDMAKKETVWRLEEFGRFASFEAQGALANIAVDKANLEIMTKRSNYTPITNVPPEVTVLTNSPVELREPNVLICFIDKFTPPVVNVTWLRNGKPVTTGVSETVFLPREDHLFRKFHYLPFLPSTEDVYDCRVEHWGLDEPLLKHWEFDAPSPLPETTENVVCALGLTVGLVGIIIGTIFIIKGVRKSNAAERRGPL'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "305c2c29",
   "metadata": {},
   "outputs": [],
   "source": [
    "s3 = 'MAVMAPRTLLLLLSGALALTQTWAGSHSMRYFFTSVSRPGRGEPRFIAVGYVDDTQFVRFDSDAASQKMEPRAPWIEQEGPEYWDQETRNMKAHSQTDRANLGTLRGYYNQSEDGSHTIQIMYGCDVGPDGRFLRGYRQDAYDGKDYIALNEDLRSWTAADMAAQITKRKWEAVHAAEQRRVYLEGRCVDGLRRYLENGKETLQRTDPPKTHMTHHPISDHEATLRCWALGFYPAEITLTWQRDGEDQTQDTELVETRPAGDGTFQKWAAVVVPSGEEQRYTCHVQHEGLPKPLTLRWELSSQPTIPIVGIIAGLVLLGAVITGAVVAAVMWRRKSSDRKGGSYTQAASSDSAQGSDVSLTACKV'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "5668a7f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YAFFQFSGGAILNTL\n"
     ]
    }
   ],
   "source": [
    "positions = [39, 41, 52, 54, 61, 82, 83, 88, 89, 91, 95, 96, 98, 102, 103 ]\n",
    "\n",
    "site = ''\n",
    "for p in positions:\n",
    "    site += s2[p]\n",
    "print(site)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "383501a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'F'"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s2[30+22]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "d5a2bc27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "82"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s1.find('FS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "98a2938c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'F'"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s1[82]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "esm_kernel",
   "language": "python",
   "name": "esm"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
